docs/proposal-platforms.md

```diff
@@ -0,0 +1,259 @@
+# Proposal - initial support for platforms
+
+This will close [cylc-flow #3421](https://github.com/cylc/cylc-flow/issues/3421)
+and partially address
+[cylc-flow #2199](https://github.com/cylc/cylc-flow/issues/2199).
+
+## Key functionality
+
+* Introduce configuration and logic to recognise compute platforms in place of
+  task hosts. This will resolve a number of issues.
+  * In particular, this allows a compute cluster which can be accessed via
+    multiple login nodes (hosts) to be treated as logical unit.
+  * We will be able to configure multiple platforms which share the same hosts
+    (including `localhost`).
+
+* If a platform has a list of hosts then the host used for each operation on that
+  platform (job submission, polling, log retrieval, etc) will be randomised. This
+  should help balance the load. If a host becomes unavailable, try another random
+  host from the list, until exhausted. This will handle hosts becoming
+  unavailable after a job is submitted which is a major limitation of the current
+  logic.
+
+* It will be possible to specify the platform for a task as a list. For example,
+  you may have 2 different HPC systems and you are happy for a task to run on
+  either. The list will be randomised and the platforms tried in turn until
+  successful job submission.
+  * Note that the platform will be selected each time a job is run. If you
+    specify a list of platforms, tasks must not assume that retries of that task
+    will use the same platform.
+  * For convenience it will also be possible to specify a platform alias which
+    defines a list of platforms.
+
+* It will be possible to specify the platform as a command in the same way as we
+  can currently specify hosts as a command. The command must return a
+  valid platform (or a list of valid platforms). We hope that this functionality
+  will seldom be needed in the future (the main use at present is to select from
+  a list of login nodes). However, there are still cases where this may be
+  needed. For example running a command which:
+  * Returns the platform which has the shortest queue time.
+  * Chooses the platform according to the day of the week (perhaps you have a
+    platform which can only be used at weekends?).
+
+* Note: it will **not** be possible to specify the platform as an environment
+  variable in the same way as we can currently specify remote hosts as an
+  environment variable (e.g. `platform = $ROSE_ORIG_HOST`).
+  * Unclear why we need this given the Jinja2 variables can be used?
+  * Using `$ROSE_ORIG_HOST` relies on Rose specifying extra options to set this
+    in `[cylc][environment]` which we should avoid if possible.
+
+## Example platform configurations
+
+```ini
+[platforms]
+    [[desktop\d\d|laptop\d\d]]
```

Comment:
Using regexes in this way presents problems for inheritance / overriding due to the multitude of ways in which a regex could be specified.

Will need to think on whether there is a nice way to do this better.